scala-thoughts.md

```diff
@@ -0,0 +1,41 @@
+
+Scala is an incredibly powerful language that I really enjoyed using through 
+these exercises. Although I have worked with quite a bit of Scala code before
+this assignment, I haven't had the opportunity to write much of it from 
+scratch. Additionally, I haven't worked with Scala much since completing PLs
+last semester. Some of the "quirky" nature of Scala makes much more sense after
+having implemented a type checker and made extensive use of Haskell in CS131.
+
+For instance, although the rules for verboseness seemed arbitrary the first
+time I worked with Scala, I now have a much better idea of when I need '()' or
+a '.', etc. This comes from understanding how a type checker would attempt to
+infer types. Additionally the value of "no side effects" makes a lot more sense
+after coming off of using the ultimate "no side effects" language, Haskell.
+
+I personally think Scala is better than Haskell (for at least
+my purposes). This is because Scala, by also supporting Object Oriented
+Programming, allows convenient solutions that Haskell needed elaborate work
+arounds for. For instance, although the Haskell designers found a clever way
+to allow printing through Monads without violating fundamental principles of
+the language, this felt more like a "hack" to allow necessary functionality.
+In some respects, although lazy evaluation and no side effects has some
+wonderful benefits, Haskell's practicality feels hampered in some ways by its
+mathematical beauty. Scala, from what I can tell, doesn't suffer from this
+since it takes the middle ground on all aspects of Object Oriented Programming
+and Functional Programming.
+
+That being said, Scala seems to also have its own short comings. I've never
+felt a programming language had as much relevance as Scala to the words
+of Spiderman's Uncle Ben, "With great power comes great responsibility!" This
+is because Scala gives the programmer a ton of freedom. But this freedom comes
+with a heavy burden the programmer must be carry. Since it feels like there
+is almost 10 ways to write just about everything, that means there is probably
+9 ways to write something bad and 1 way to write it good. This means keeping
+to standard practices is very important to keep your code readable in Scala.
+
```

Comment:
Maybe so, but I think that eventually we'll be able to differentiate which is best for each situation. I do agree that it's burdensome, though. Often I feel like I did something ever so slightly wrong and end up trying to fix it by putting types everywhere...